README.md
# logfmtr

An implementation of the [logr minimal logging API](https://github.com/go-logr/logr) that writes in [logfmt](https://www.brandur.org/logfmt) style.

[![Build Status](https://travis-ci.org/iand/logfmtr.svg?branch=master)](https://travis-ci.org/iand/logfmtr)
[![Go Report Card](https://goreportcard.com/badge/github.com/iand/logfmtr)](https://goreportcard.com/report/github.com/iand/logfmtr)
[![go.dev reference](https://img.shields.io/badge/go.dev-reference-007d9c?logo=go&logoColor=white)](https://pkg.go.dev/github.com/iand/logfmtr)

## Getting Started

This is a no frills logging package that follows the [logr minimal logging API](https://github.com/go-logr/logr) 
and by default writes logs in [logfmt](https://www.brandur.org/logfmt) format, a line of space delimited
key/value pairs.

```
level=0 logger=MyName ts=2020-09-20T14:31:10.905267839Z msg=hello user=you val1=1 val2=map[k:1]
level=1 logger=MyName ts=2020-09-20T14:31:10.905279546Z msg="you should see this" user=you
level=0 logger=MyName ts=2020-09-20T14:31:10.905288008Z msg="uh oh" error=<nil> user=you trouble=true reasons="[0.1 0.11 3.14]"
level=0 logger=MyName ts=2020-09-20T14:31:10.905291479Z msg=goodbye error="an error occurred" user=you code=-1
```

A more human friendly output format is also provided which can be configured using the `humanize` option:

```
0 info  | 14:31:10.905297 | hello                          logger=MyName user=you val1=1 val2=map[k:1]
1 info  | 14:31:10.905302 | you should see this            logger=MyName user=you
0 error | 14:31:10.905307 | uh oh                          logger=MyName error=<nil> user=you trouble=true reasons="[0.1 0.11 3.14]"
0 error | 14:31:10.905311 | goodbye                        logger=MyName error="an error occurred" user=you code=-1
```

Loggers defer applying their configuration until they are used. The logger is instantiated when
either Info, Error or Enabled is called. At that point the logger will read and use any options set
from a prior call to UseOptions. 


```Go
package main

import (
    "github.com/iand/logfmtr"
)

func main() {
    // Set options that all loggers will be based on
    opts := logfmtr.DefaultOptions()
    opts.Humanize = true
    opts.AddCaller = true
    logfmtr.UseOptions(opts)
}

```

Any new loggers will use the options set in `main` when they first start logging.

```Go
package worker

import (
    "github.com/iand/logfmtr"
)

// Create the logger
var logger = logfmtr.New().WithName("worker")


func doWork() }
    // Logger is instantiated with the options set earlier
    logger.Info("the sun is shining")
}
```

Loggers can be created with specific configuration by using `NewWithOptions`:

```Go
func otherWork() }
    opts := logfmtr.DefaultOptions()
    opts.Writer = os.Stderr

    logger := logfmtr.NewWithOptions(opts)
    logger.Info("important system information")
}
```

Several predefined keys are used when writing logs in logfmt style:

 * **msg** - the log message
 * **error** - error message passed to the `Error` method
 * **logger** - the name of the logger writing the log entry
 * **level** - the verbosity level of the logger writing the log entry
 * **caller** - filename and line number of the origin of the log entry

## Author

* [Ian Davis](http://github.com/iand) - <http://iandavis.com/>

## License

This is free and unencumbered software released into the public domain. Anyone is free to 
copy, modify, publish, use, compile, sell, or distribute this software, either in source 
code form or as a compiled binary, for any purpose, commercial or non-commercial, and by 
any means. For more information, see <http://unlicense.org/> or the 
accompanying [`UNLICENSE`](UNLICENSE) file.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

These are changes that will probably be included in the next release.

### Added
 * Add benchmarks

## [v0.2.1] - 2021-09-01

### Fixed
 * Fixed panic when Error is first call on a logger (thanks @wutz for report)
 * Updated readme example to match actual output (thanks @seankhliao)


## [v0.2.0] - 2021-07-07

This version supports the logr v1.0 API which contains breaking changes from pre-1.0 versions.

### Changed
 * Support logr v1.0 API 
 * Write ts field before msg in standard output.
 * Use a fixed number of fractional seconds in timestamp.
 
## [v0.1.5] - 2020-11-23

### Added
 * Add DisableLogger and EnableLogger to disable/enable specific named loggers.

### Removed
 * Deprecated Null logger and it will be removed in next version, use logr.Discard instead
 * Deprecated FromContext and NewContext and they will be removed in next version, use logr.FromContext and logr.NewContext 

## [v0.1.4] - 2020-10-11

### Added
 * Add Null logger
 * Add NewNamed convenience function
 * Add Context support

## [v0.1.3] - 2020-09-20

### Changed 
 * Change to deferred loggers that instantiate their configuration on first use.

## [v0.1.2] - 2020-09-13

### Changed 
 * Updated travis configuration to target recent Go versions

## [v0.1.1] - 2020-07-31

### Added
 * New AddCaller and CallerSkip options to add caller information to log messages

## [v0.1.0] - 2020-07-31

Initial release


logfmtr_test.go
package logfmtr_test

import (
	"io"
	"testing"

	"github.com/iand/logfmtr"
)

func ExampleNewWithOptions() {
	opts := logfmtr.DefaultOptions()
	opts.Humanize = true
	logger := logfmtr.NewWithOptions(opts)
	logger.Info("the sun is shining")
}

func ExampleNew() {
	// Create the logger with deferred options
	logger := logfmtr.New()

	// Later on set options that all loggers will be based on
	opts := logfmtr.DefaultOptions()
	opts.Humanize = true
	logfmtr.UseOptions(opts)

	// Logger is instantiated with the options
	logger.Info("the sun is shining")
}

func ExampleNewNamed() {
	// Create the logger with a name
	logger := logfmtr.NewNamed("europa")

	logger.Info("the sun is shining")
}

func ExampleDisableLogger() {
	// Create the logger with a name
	logger := logfmtr.NewNamed("europa")

	// It logs normally
	logger.Info("the sun is shining")

	// Disable the logger
	logfmtr.DisableLogger("europa")

	logger.Info("the sun not shining now")
}

func discard() logfmtr.Options {
	opts := logfmtr.DefaultOptions()
	opts.Writer = io.Discard
	return opts
}

func TestIssue3(t *testing.T) {
	defer func() {
		if r := recover(); r != nil {
			t.Fatalf("unexpected panic: %v", r)
		}
	}()
	logfmtr.SetVerbosity(1)
	base := logfmtr.NewWithOptions(discard())

	log := base.WithValues("user", "you")

	// Should not panic
	log.Error(nil, "uh oh", "trouble", true, "reasons", []float64{0.1, 0.11, 3.14})
}


benchmark_test.go
package logfmtr_test

import (
	"fmt"
	"testing"

	"github.com/go-logr/logr"
	"github.com/iand/logfmtr"
)

// Benchmarks copied from github.com/go-logr/logr/benchmark

//go:noinline
func doInfoOneArg(b *testing.B, log logr.Logger) {
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		log.Info("this is", "a", "string")
	}
}

//go:noinline
func doInfoSeveralArgs(b *testing.B, log logr.Logger) {
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		log.Info("multi",
			"bool", true, "string", "str", "int", 42,
			"float", 3.14, "struct", struct{ X, Y int }{93, 76})
	}
}

//go:noinline
func doV0Info(b *testing.B, log logr.Logger) {
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		log.V(0).Info("multi",
			"bool", true, "string", "str", "int", 42,
			"float", 3.14, "struct", struct{ X, Y int }{93, 76})
	}
}

//go:noinline
func doV9Info(b *testing.B, log logr.Logger) {
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		log.V(9).Info("multi",
			"bool", true, "string", "str", "int", 42,
			"float", 3.14, "struct", struct{ X, Y int }{93, 76})
	}
}

//go:noinline
func doError(b *testing.B, log logr.Logger) {
	b.ReportAllocs()
	b.ResetTimer()
	err := fmt.Errorf("error message")
	for i := 0; i < b.N; i++ {
		log.Error(err, "multi",
			"bool", true, "string", "str", "int", 42,
			"float", 3.14, "struct", struct{ X, Y int }{93, 76})
	}
}

//go:noinline
func doWithValues(b *testing.B, log logr.Logger) {
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		l := log.WithValues("k1", "v1", "k2", "v2")
		_ = l
	}
}

//go:noinline
func doWithName(b *testing.B, log logr.Logger) {
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		l := log.WithName("name")
		_ = l
	}
}

//go:noinline
func doWithCallDepth(b *testing.B, log logr.Logger) {
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		l := log.WithCallDepth(1)
		_ = l
	}
}

func newLogger() logr.Logger {
	return logfmtr.NewWithOptions(discard())
}

func BenchmarkLogfmtrInfoOneArg(b *testing.B) {
	doInfoOneArg(b, newLogger())
}

func BenchmarkLogfmtrInfoSeveralArgs(b *testing.B) {
	doInfoSeveralArgs(b, newLogger())
}

func BenchmarkLogfmtrV0Info(b *testing.B) {
	doV0Info(b, newLogger())
}

func BenchmarkLogfmtrV9Info(b *testing.B) {
	doV9Info(b, newLogger())
}

func BenchmarkLogfmtrError(b *testing.B) {
	doError(b, newLogger())
}

func BenchmarkLogfmtrWithValues(b *testing.B) {
	doWithValues(b, newLogger())
}

func BenchmarkLogfmtrWithName(b *testing.B) {
	doWithName(b, newLogger())
}

func BenchmarkLogfmtrWithCallDepth(b *testing.B) {
	doWithCallDepth(b, newLogger())
}


logfmtr.go
package logfmtr

import (
	"bytes"
	"fmt"
	"io"
	"os"
	"path"
	"runtime"
	"strconv"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"github.com/go-logr/logr"
)

// The global verbosity level.
var gv int32 = 0

// SetVerbosity sets the global log level. Only loggers with a V level less than
// or equal to this value will be enabled.
func SetVerbosity(v int) int {
	old := atomic.SwapInt32(&gv, int32(v))
	return int(old)
}

var (
	goptionsmu sync.Mutex
	goptions   = DefaultOptions()

	disabledLoggersMu sync.Mutex // synchronises writes to disabledLoggers map
	disabledLoggers   atomic.Value
	anyDisabled       int32 = 0 // an atomicly accessed variable that is set to 1 if there are any loggers that have been manually disabled
)

func init() {
	disabledLoggers.Store(map[string]bool{})
}

// UseOptions sets options that new loggers will use when it they are instantiated.
func UseOptions(opts Options) {
	goptionsmu.Lock()
	goptions = opts
	goptionsmu.Unlock()
}

// New returns a deferred logger that writes in logfmt using the default options.
// The logger defers configuring its options until it is instantiated with the first call to Info, Error
// or Enabled or the first call to those function on any child loggers created via V, WithName or
// WithValues.
func New() logr.Logger {
	return logr.New(&sink{})
}

// NewNamed returns a deferred logger with the given name that writes in logfmt using the default options.
func NewNamed(name string) logr.Logger {
	return New().WithName(name)
}

// NewWithOptions returns an instantiated logger that writes in logfmt using the supplied options. Panics if
// no writer is supplied in the options.
func NewWithOptions(opts Options) logr.Logger {
	s := &sink{}
	s.applyOptions(opts)
	return logr.New(s)
}

// DefaultOptions returns the default options used by New unless overridden by a call to UseOptions.
// Override the option field to customise behaviour and then pass to UseOptions or NewWithOptions.
func DefaultOptions() Options {
	return Options{
		Writer:          os.Stdout,
		TimestampFormat: "2006-01-02T15:04:05.000000000Z07:00",
		NameDelim:       ".",
	}
}

// Options contains fields and flags for customizing logger behaviour
type Options struct {
	// Writer is where logs will be written to
	Writer io.Writer

	// Humanize changes the log output to a human friendly format
	Humanize bool

	// Colorize adds color to the log output. Only applies if Humanize is also true.
	Colorize bool

	// TimestampFormat sets the format for log timestamps. Set to empty to disable timestamping
	// of log messages. Humanize uses a fixed short timestamp format.
	TimestampFormat string

	// NameDelim is the delimiter character used when appending names of loggers.
	NameDelim string

	// AddCaller indicates that log messages should include the file and line number of the caller of the logger.
	AddCaller bool

	// CallerSkip adds frames to skip when determining the caller of the logger. Useful when the logger is wrapped
	// by another logger.
	CallerSkip int
}

var _ logr.LogSink = (*sink)(nil)

// sink is a logger sink that writes messages in the logfmt style.
// See https://www.brandur.org/logfmt for more information.
type sink struct {
	core        *core
	init        sync.Once
	parent      *sink
	runtimeInfo logr.RuntimeInfo
	dfn         func(*core)
}

func (l *sink) instantiate() {
	if l.core != nil {
		return
	}
	if l.parent == nil {
		goptionsmu.Lock()
		l.core = &core{
			runtimeInfo: l.runtimeInfo,
		}
		l.core.applyOptions(goptions)
		goptionsmu.Unlock()
		if l.dfn != nil {
			l.dfn(l.core)
		}
		return
	}

	l.core = l.parent.copyCore(l.dfn)
}

func (l *sink) copyCore(dfn func(*core)) *core {
	l.init.Do(l.instantiate)

	c := *l.core
	dfn(&c)
	return &c
}

func (l *sink) applyOptions(opts Options) {
	l.core = &core{}
	l.core.applyOptions(opts)
}

func (l *sink) Init(info logr.RuntimeInfo) {
	l.runtimeInfo = info
}

// Enabled reports whether this Logger is enabled with respect to the current global log level.
func (l *sink) Enabled(level int) bool {
	l.init.Do(l.instantiate)
	if level > int(atomic.LoadInt32(&gv)) {
		return false
	}
	if l.core.name == "" || atomic.LoadInt32(&anyDisabled) == 0 {
		return true
	}
	disabled := disabledLoggers.Load().(map[string]bool)
	return !disabled[l.core.name]
}

// Info logs a non-error message with the given key/value pairs as context.
func (l *sink) Info(level int, msg string, kvs ...interface{}) {
	l.init.Do(l.instantiate)
	l.core.write(level, "info", msg, l.core.flatten(kvs...))
}

// Error logs an error, with the given message and key/value pairs as context.
func (l *sink) Error(err error, msg string, kvs ...interface{}) {
	l.init.Do(l.instantiate)
	l.core.write(0, "error", msg, l.core.flatten(kvs...), "error", err)
}

// WithName returns a logger with a new element added to the logger's name.
func (l *sink) WithName(name string) logr.LogSink {
	return &sink{
		parent: l,
		dfn: func(c *core) {
			c.appendName(name)
		},
	}
}

// WithValues returns a logger with additional key-value pairs of context.
func (l *sink) WithValues(kvs ...interface{}) logr.LogSink {
	return &sink{
		parent: l,
		dfn: func(c *core) {
			values := c.flatten(kvs...)
			c.appendValues(values)
		},
	}
}

func (l *sink) WithCallDepth(depth int) logr.LogSink {
	return &sink{
		parent: l,
		dfn: func(c *core) {
			c.callerSkip += depth
		},
	}
}

type core struct {
	w           io.Writer
	name        string
	values      string
	humanize    bool
	tsFormat    string
	nameDelim   string
	colorize    bool
	addCaller   bool
	callerSkip  int
	runtimeInfo logr.RuntimeInfo
}

func (c *core) write(level int, humanprefix, msg string, values string, extras ...interface{}) {
	var b bytes.Buffer
	if c.humanize {
		if c.colorize {
			if humanprefix == "error" {
				humanprefix = colorRed + humanprefix + colorDefault
			} else {
				humanprefix = colorGreen + humanprefix + " " + colorDefault
			}
		}

		b.WriteString(fmt.Sprintf("%d %-5s | %15s | %-30s", level, humanprefix, time.Now().UTC().Format("15:04:05.000000"), msg))
		if c.name != "" {
			b.WriteRune(' ')
			b.WriteString(c.key("logger"))
			b.WriteString("=")
			b.WriteString(c.name)
		}
		if c.addCaller {
			b.WriteRune(' ')
			b.WriteString(c.key("caller"))
			b.WriteString("=")
			b.WriteString(c.caller(1))
		}
	} else {
		b.WriteString("level=")
		b.WriteString(strconv.Itoa(level))
		if c.name != "" {
			b.WriteRune(' ')
			b.WriteString("logger=")
			b.WriteString(quote(c.name))
		}
		if c.tsFormat != "" {
			b.WriteRune(' ')
			b.WriteString("ts=")
			b.WriteString(quote(time.Now().UTC().Format(c.tsFormat)))
		}
		b.WriteRune(' ')
		b.WriteString("msg=")
		b.WriteString(quote(msg))
		if c.addCaller {
			b.WriteRune(' ')
			b.WriteString("caller=")
			b.WriteString(c.caller(1))
		}
	}
	if len(extras) > 0 {
		b.WriteRune(' ')
		b.WriteString(c.flatten(extras...))
	}

	if c.values != "" {
		b.WriteRune(' ')
		b.WriteString(c.values)
	}
	if values != "" {
		b.WriteRune(' ')
		b.WriteString(values)
	}
	b.WriteRune('\n')
	_, _ = c.w.Write(b.Bytes())
}

func (c *core) caller(skip int) string {
	for i := 1; i < 3; i++ {
		_, file, line, ok := runtime.Caller(c.runtimeInfo.CallDepth + skip + c.callerSkip + i)
		if ok && file != "<autogenerated>" {
			return path.Base(file) + ":" + strconv.Itoa(line)
		}
	}
	return "unknown"
}

func (c *core) applyOptions(opts Options) {
	if opts.Writer == nil {
		panic("logger was supplied with nil writer")
	}
	c.w = opts.Writer
	c.humanize = opts.Humanize
	c.tsFormat = opts.TimestampFormat
	c.nameDelim = opts.NameDelim
	c.colorize = opts.Colorize && opts.Humanize
	c.addCaller = opts.AddCaller
	c.callerSkip = opts.CallerSkip
}

func (c *core) flatten(kvs ...interface{}) string {
	if len(kvs) == 0 {
		return ""
	}
	var b strings.Builder
	for i := 0; i < len(kvs); i += 2 {
		if i > 0 {
			b.WriteRune(' ')
		}

		k := kvs[i]
		var v interface{}
		if i+1 < len(kvs) {
			v = kvs[i+1]
		} else {
			v = ""
		}
		b.WriteString(c.key(stringify(k)))
		b.WriteRune('=')
		b.WriteString(stringify(v))
	}

	return b.String()
}

func (c *core) key(s string) string {
	if !c.colorize {
		return s
	}

	switch s {
	case "error":
		return colorRed + s + colorDefault
	case "logger", "caller":
		return colorBlue + s + colorDefault
	default:
		return colorYellow + s + colorDefault
	}
}

func (c *core) appendName(name string) {
	if name == "" {
		return
	}
	if c.name != "" {
		c.name = c.name + c.nameDelim + name
	} else {
		c.name = name
	}
}

func (c *core) appendValues(values string) {
	if values == "" {
		return
	}
	if len(c.values) > 0 {
		c.values = c.values + " " + values
	} else {
		c.values = values
	}
}

func stringify(v interface{}) string {
	var s string
	switch vv := v.(type) {
	case string:
		s = vv
	case fmt.Stringer:
		s = vv.String()
	case error:
		s = vv.Error()
	default:
		s = fmt.Sprint(v)
	}
	return quote(s)
}

func quote(s string) string {
	if strings.ContainsAny(s, " ") {
		return fmt.Sprintf("%q", s)
	}
	return s
}

const (
	colorDefault = "\x1b[0m"
	colorRed     = "\x1b[1;31m"
	colorGreen   = "\x1b[1;32m"
	colorYellow  = "\x1b[1;33m"
	colorBlue    = "\x1b[1;34m"
)

func DisableLogger(name string) {
	setLoggerDisabledStatus(name, true)
}

func EnableLogger(name string) {
	setLoggerDisabledStatus(name, false)
}

func setLoggerDisabledStatus(name string, disabled bool) {
	disabledLoggersMu.Lock()
	defer disabledLoggersMu.Unlock()
	current := disabledLoggers.Load().(map[string]bool)
	next := make(map[string]bool, len(current))
	for k, v := range current {
		if k == name {
			continue
		}
		next[k] = v
	}
	if disabled {
		next[name] = disabled
	}
	disabledLoggers.Store(next)
	if len(next) == 0 {
		atomic.StoreInt32(&anyDisabled, 0)
	} else {
		atomic.StoreInt32(&anyDisabled, 1)
	}
}


examples/example.go
package main

import (
	"fmt"

	"github.com/go-logr/logr"
	"github.com/iand/logfmtr"
)

func main() {
	logfmtr.SetVerbosity(1)

	demo(logfmtr.New())

	opts := logfmtr.DefaultOptions()
	opts.Humanize = true
	opts.Colorize = true
	opts.CallerSkip = 0
	demo(logfmtr.NewWithOptions(opts))

	deferred()

	disableDemo()
}

func demo(base logr.Logger) {
	log := base.WithName("MyName").WithValues("user", "you")
	log.Info("hello", "val1", 1, "val2", map[string]int{"k": 1})
	log.V(1).Info("you should see this")
	log.V(1).V(1).Info("you should NOT see this")
	log.Error(nil, "uh oh", "trouble", true, "reasons", []float64{0.1, 0.11, 3.14})
	log.Error(fmt.Errorf("an error occurred"), "goodbye", "code", -1)
}

func deferred() {
	l1 := logfmtr.New().WithName("before")
	l2 := logfmtr.NewNamed("after")
	l3 := l2.WithValues("some", "value")

	l1.Info("this should be logged with default options, not humanized or colorized")
	opts := logfmtr.DefaultOptions()
	opts.Humanize = true
	opts.Colorize = true
	logfmtr.UseOptions(opts)

	l2.Info("this should be logged with new global options, humanized and colorized, since instantiation was deferred until first write")
	l3.Info("this should also be logged with new options, humanized and colorized")
	l1.Info("this should be logged with the old options since first write was before we set global options")
}

func disableDemo() {
	log := logfmtr.NewNamed("europa")
	log.Info("hello, this logger is enabled")

	logfmtr.DisableLogger("europa")
	log.Info("you should NOT see this, the logger is disabled")

	log2 := log.WithName("moon")
	log2.Info("you should see this, a child logger does not inherit from its parent")

	logfmtr.EnableLogger("europa")
	log.Info("you should see this now, the logger was enabled again")
}
